packages/ckeditor5-utils/src/dom/position.js

```diff
@@ -310,56 +238,228 @@ function getBestOfProcessedPositions( processedPositions, elementRectArea ) {
 //
 // @private
 // @param {utils/dom/rect~Rect} rect A rect to be converted.
+//
 // @returns {Object} Object containing `left` and `top` properties, in absolute coordinates.
-function getAbsoluteRectCoordinates( { left, top } ) {
+function getRectForAbsolutePositioning( rect ) {
 	const { scrollX, scrollY } = global.window;
 
-	return {
-		left: left + scrollX,
-		top: top + scrollY
-	};
+	return rect.clone().moveBy( scrollX, scrollY );
+}
+
+/**
+ * Position class used to store and calculate various positioning components used for example in balloons positioning.
+ *
+ * Usage example:
+ *
+ *		const elementRect = new Rect( element );
+ *		const targetRect = new Rect( target );
+ *		const viewportRect = new Rect( viewport );
+ *		const positionedElementAncestor = getPositionedAncestor( element );
+ *
+ *		const options = { elementRect, elementRect, viewportRect, positionedElementAncestor };
+ *
+ *		const positioningFunction = ( element, target, viewport ) => ( { top: 0, left: 0, name: 'example', withArrow: false } );
+ *
+ *		const position = new Position( positioningFunction, options );
+ *
+ *		const { top, left, name, withArrow, limiterIntersectionArea, viewportIntersectionArea } = position;
+ *
+ * The above `positioningFunction` is only an example always returning hardcoded top and left coordinates regardless of given options.
+ *
+ * Positioning function must be compatible with {@link module:utils/dom/position~Position}.
+ *
+ * Check the {@link module:ui/panel/balloon/balloonpanelview~BalloonPanelView.defaultPositions `BalloonPanelView.defaultPositions`}
+ * for example functions.
+ */
+class Position {
+	/**
+	 * Creates an instance of the {@link module:utils/dom/position~Position} class.
+	 *
+	 * @param {Function} [positioningFunction] function handling rects positioning.
+	 * @param {Object} [options] options object.
+	 * @param {module:utils/dom/rect~Rect} [options.targetRect] target element rect.
+	 * @param {module:utils/dom/rect~Rect} [options.elementRect] balloon element rect.
+	 * @param {module:utils/dom/rect~Rect} [options.viewportRect] viewport element rect.
+	 * @param {HTMLElement|null} [options.positionedElementAncestor] nearest ancestor element which CSS position is not "static".
+	 */
+	constructor( positioningFunction, options ) {
+		const positioningFunctionOutput = positioningFunction( options.targetRect, options.elementRect, options.viewportRect );
+
+		// Nameless position for a function that didn't participate.
+		if ( !positioningFunctionOutput ) {
+			return;
+		}
+
+		const { left, top, name, withArrow } = positioningFunctionOutput;
+
+		Object.assign( this, { name, withArrow } );
+
+		this._positioningFunctionCorrdinates = { left, top };
+		this._options = options;
+
+		/**
+		 * Position name.
+		 *
+		 * @readonly
+		 * @member {String} #name
+		 */
+
+		/**
+		 * Tells whether given balloon constructed out of this position should have arrow.
+		 *
+		 * @readonly
+		 * @member {String} #withArrow
+		 */
```

Comment:
Having this property here mixes the positioning logic and feature logic (the former should be feature-agnostic).

This property is relevant for `BalloonPanelView`. However, other use-cases such as `DropdownView` don't have such thing (dropdowns don't have arrows). Also, there could be other use-cases that have nothing to do with arrows and balloons: `getOptimalPosition()` should position virtually anything.

So instead of `#withArrow` let's have `#config` property (or `#presentation` which is more specific) that will pass through any configuration from the positioning function like a payload. For instance, the positioning function will be

```
	viewportStickyNorth: ( targetRect, balloonRect, viewportRect ) => {
		if ( !targetRect.getIntersection( viewportRect ) ) {
			return null;
		}

		return {
			top: viewportRect.top + BalloonPanelView.stickyVerticalOffset,
			left: targetRect.left + targetRect.width / 2 - balloonRect.width / 2,
			name: 'arrowless',
			config: {
				withArrow: false
			}			
		};
	}
```

and the config will land in `Position#config` and `BalloonPanelView` having received such an optimal position will know what to do with its config (for instance, default it to `{ withArrow: true }`).